heartbeat/scheduler/scheduler.go

```diff
@@ -139,281 +182,119 @@ var ErrAlreadyStopped = errors.New("attempted to add job to already stopped sche
 // Add adds the given TaskFunc to the current scheduler. Will return an error if the scheduler
 // is done.
 func (s *Scheduler) Add(sched Schedule, id string, entrypoint TaskFunc) (removeFn func() error, err error) {
-	debugf("Add scheduler job '%v'.", id)
-
-	j := &job{
-		id:         id,
-		fn:         entrypoint,
-		schedule:   sched,
-		registered: false,
-		running:    0,
-	}
-	if s.isPreRunning() {
-		s.addSync(j)
-	} else if s.isRunning() {
-		s.addCh <- j
-	} else {
+	if s.state.Load() == stateDone {
 		return nil, ErrAlreadyStopped
 	}
 
-	return func() error { return s.remove(j) }, nil
-}
-
-func (s *Scheduler) remove(j *job) error {
-	debugf("Remove scheduler job '%v'", j.id)
-
-	if s.isPreRunning() {
-		s.doRemove(j)
-	} else if s.isRunning() {
-		s.rmCh <- j
-	}
-	// There is no need to handle the isDone case
-	// because removing the job accomplishes nothing if
-	// the scheduler is stopped
-
-	return nil
-}
-
-func (s *Scheduler) run() {
-	defer func() {
-		// drain finished queue for active jobs to not leak
-		// go-routines on exit
-		for i := uint(0); i < s.active; i++ {
-			<-s.finished
+	removed := atomic.MakeBool(false)
+	removeCh := make(chan bool)
+	removeFn = func() error {
+		// Safely close the channel exactly once
+		// Clients may invoke this function multiple times
+		if removed.CAS(false, true) {
```

Comment:
why not `sync.Once`?

Reply:
Good call, I forgot that existed! Will fix.